pkg/ddc/juicefs/sync_runtime.go

```diff
@@ -152,3 +250,155 @@ func (j *JuiceFSEngine) syncFuseSpec(ctx cruntime.ReconcileRequestContext, runti
 	}
 	return
 }
+
+func (j *JuiceFSEngine) isVolumeMountsChanged(crtVolumeMounts, runtimeVolumeMounts []corev1.VolumeMount) (changed bool, newVolumeMounts []corev1.VolumeMount) {
+	mounts := make(map[string]corev1.VolumeMount)
+	for _, mount := range crtVolumeMounts {
+		mounts[mount.Name] = mount
+	}
+	for _, mount := range runtimeVolumeMounts {
+		if m, ok := mounts[mount.Name]; !ok || !reflect.DeepEqual(m, mount) {
+			j.Log.Info("The volumeMounts is different.", "current sts", crtVolumeMounts, "runtime", runtimeVolumeMounts)
+			mounts[mount.Name] = mount
+			changed = true
+		}
+	}
+	vms := []corev1.VolumeMount{}
+	for _, mount := range mounts {
+		vms = append(vms, mount)
+	}
+	newVolumeMounts = vms
+	return
+}
+
+func (j JuiceFSEngine) isEnvsChanged(crtEnvs, runtimeEnvs []corev1.EnvVar) (changed bool, newEnvs []corev1.EnvVar) {
+	envMap := make(map[string]string)
+	for _, env := range crtEnvs {
+		envMap[env.Name] = env.Value
+	}
+	for _, env := range runtimeEnvs {
+		if envMap[env.Name] != env.Value {
+			j.Log.Info("The env is different.", "current sts", crtEnvs, "runtime", runtimeEnvs)
+			envMap[env.Name] = env.Value
+			changed = true
+		}
+	}
+	envs := []corev1.EnvVar{}
+	for name, env := range envMap {
+		envs = append(envs, corev1.EnvVar{Name: name, Value: env})
+	}
+	newEnvs = envs
+	return
+}
+
+func (j JuiceFSEngine) isResourcesChanged(crtResources, runtimeResources corev1.ResourceRequirements) (changed bool, newResources corev1.ResourceRequirements) {
+	if !utils.ResourceRequirementsEqual(crtResources, runtimeResources) {
+		j.Log.Info("The resource requirement is different.", "current sts", crtResources, "runtime", runtimeResources)
+		changed = true
+	}
+	newResources = runtimeResources
+	return
+}
+
+func (j JuiceFSEngine) isVolumesChanged(crtVolumes, runtimeVolumes []corev1.Volume) (changed bool, newVolumes []corev1.Volume) {
+	volumes := make(map[string]corev1.Volume)
+	for _, mount := range crtVolumes {
+		volumes[mount.Name] = mount
+	}
+	for _, volume := range runtimeVolumes {
+		if m, ok := volumes[volume.Name]; !ok || !reflect.DeepEqual(m, volume) {
+			j.Log.Info("The volumes is different.", "current sts", crtVolumes, "runtime", runtimeVolumes)
+			volumes[volume.Name] = volume
+			changed = true
+		}
+	}
+	vs := []corev1.Volume{}
+	for _, volume := range volumes {
+		vs = append(vs, volume)
+	}
+	newVolumes = vs
+	return
+}
```

Comment:
What if `runtimeVolumes` deletes some volume from `crtVolumes`? The func now takes such situation as "unchanged".